dialog/file.go

```diff
@@ -445,44 +448,63 @@ func (f *fileDialog) setLocation(dir fyne.URI) error {
 	return nil
 }
 
-func (f *fileDialog) setSelected(file *fileDialogItem) {
-	if f.selected != nil {
-		f.selected.isCurrent = false
-		f.selected.Refresh()
-	}
-	if file != nil && file.isDirectory() {
-		listable, err := storage.CanList(file.location)
-		if err != nil || !listable {
-			fyne.LogError("Failed to create lister for URI"+file.location.String(), err)
+func (f *fileDialog) setSelected(file fyne.URI, id int) {
+	if file != nil {
+		if listable, err := storage.CanList(file); err == nil && listable {
+			f.setLocation(file)
+			return
 		}
-		f.setLocation(file.location)
-		return
 	}
 	f.selected = file
+	f.selectedID = id
 
-	if file == nil || file.location.String()[len(file.location.Scheme())+3:] == "" {
+	if file == nil || file.String()[len(file.Scheme())+3:] == "" {
 		// keep user input while navigating
 		// in a FileSave dialog
 		if !f.file.save {
 			f.fileName.SetText("")
 			f.open.Disable()
 		}
 	} else {
-		file.isCurrent = true
-		f.fileName.SetText(file.location.Name())
+		f.fileName.SetText(file.Name())
 		f.open.Enable()
 	}
 }
 
 func (f *fileDialog) setView(view viewLayout) {
 	f.view = view
+	count := func() int {
+		return len(f.data)
+	}
+	template := func() fyne.CanvasObject {
+		return f.newFileItem(storage.NewFileURI("./tempfile"), true, false)
+	}
 	if f.view == gridView {
-		padding := fyne.NewSize(fileIconCellWidth-fileIconSize, theme.Padding())
-		f.files = container.NewGridWrap(
-			fyne.NewSize(fileIconSize, fileIconSize+fileTextSize).Add(padding),
+		f.files = widget.NewGridWrap(count, template,
+			func(id widget.GridWrapItemID, o fyne.CanvasObject) {
+				dir := f.data[id]
+				parent := id == 0 && len(dir.Path()) < len(f.dir.Path())
+				_, isDir := dir.(fyne.ListableURI)
+				o.(*fileDialogItem).setLocation(dir, isDir || parent, parent)
+			},
 		)
+		f.files.(*widget.GridWrap).OnSelected = func(id widget.GridWrapItemID) {
+			f.selectedID = id
+			f.setSelected(f.data[id], id)
+		}
 	} else {
-		f.files = container.NewVBox()
+		f.files = widget.NewList(count, template,
+			func(id widget.ListItemID, o fyne.CanvasObject) {
+				dir := f.data[id]
+				parent := id == 0 && dir != f.dir
+				_, isDir := dir.(fyne.ListableURI)
+				o.(*fileDialogItem).setLocation(dir, isDir, parent)
+			},
+		)
+		f.files.(*widget.List).OnSelected = func(id widget.ListItemID) {
+			f.selectedID = id
+			f.setSelected(f.data[id], id)
+		}
```

Comment:
If you use a `&widget.List{}` type here, you can avoid having to cast the variable to get the `OnSelected` field and just set al of the functions directly instead.